### DISCUSSION TITLE
How is this architecture different from Uncle Bob's Clean Architecture?

### DISCUSSION THREAD

**Discussion by bladerunnersde** [LordMoMA/Hexagonal-Architecture#4] [Q&A] [answered]
I am curious with all the architectures out there, which to choose for my project?

Thank you for your thought.

**Comment by LordMoMA** [comment, marked as answer]
As in the hexagonal architecture, business logic is also at the center of clean architecture. Around it are the so-called interface adapters, which connect the core with the user interface, the database, and other external components. The core only knows the interfaces of the adapters but knows nothing about their concrete implementations and the components behind them.

In clean architecture, too, all source code dependencies point exclusively in the direction of the core. Where calls point from the inside to the outside, i.e., in the opposite direction to the source code dependency, the dependency inversion principle is applied.

The hexagons can be mapped almost one-to-one to the clean architecture rings:

1. The “external agencies” arranged around the outer hexagon correspond to the outer ring of the clean architecture, “frameworks & drivers.”
2. The outer hexagon “adapters” corresponds to the ring “interface adapters.”
3. The application hexagon corresponds to the “business rules” in the clean architecture. However, these are further subdivided into “enterprise-related business rules” (entities) and “application business rules” (use cases that orchestrate the entities and control the flow of data to and from them). On the other hand, the hexagonal architecture deliberately leaves the architecture within the application hexagon open.

Both architectures are almost identical: the software is divided into layers, and all source code dependencies point from the outer to the inner layers. The application’s core knows no details of the outer layers and is implemented only against their interfaces. This creates a system whose technical details are interchangeable and which is fully testable without them.

**Reply by bladerunnersde** [reply]
Thank you so much, this is really a thorough answer.

**Reply by LordMoMA** [reply]
> Thank you so much, this is really a thorough answer.

glad it helps ;)
